cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,497 @@
+                    cudaMemcpyDeviceToDevice, m_stream);
+
+    return toMat;
+  }
+
+  //! Visualize the first matrix.
+  void print(std::string name) const {
+    size_t len = m_shape.first * m_shape.second * m_num_batches;
+    std::vector<double> A(len);
+    updateHost(A.data(), m_A_dense.get(), len, m_stream);
+    std::cout << name << "=\n";
+    for (int i = 0; i < m_shape.first; i++) {
+      for (int j = 0; j < m_shape.second; j++) {
+        // column major
+        std::cout << std::setprecision(10) << A[j * m_shape.first + i] << ",";
+      }
+      std::cout << "\n";
+    }
+  }
+
+  /**
+   * @brief Initialize a batched identity matrix.
+   * 
+   * @param[in]  m            Number of rows/columns of matrix
+   * @param[in]  num_batches  Number of matrices in batch
+   * @param[in]  pool         Memory pool
+   * 
+   * @return A batched identity matrix
+   */
+  static BatchedMatrix Identity(int m, int num_batches, cublasHandle_t handle,
+                                std::shared_ptr<ML::deviceAllocator> allocator,
+                                cudaStream_t stream) {
+    BatchedMatrix I(m, m, num_batches, handle, allocator, stream, true);
+
+    identity_matrix_kernel<<<num_batches, std::min(1024, m), 0, stream>>>(
+      I.raw_data(), m);
+    CUDA_CHECK(cudaGetLastError());
+    return I;
+  }
+
+ private:
+  //! Shape (rows, cols) of matrices. We assume all matrices in batch have same shape.
+  std::pair<int, int> m_shape;
+
+  //! Array(pointer) to each matrix.
+  std::shared_ptr<double*> m_A_batches;
+
+  //! Data pointer to first element of dense matrix data.
+  std::shared_ptr<double> m_A_dense;
+
+  //! Number of matrices in batch
+  size_t m_num_batches;
+
+  std::shared_ptr<deviceAllocator> m_allocator;
+  cublasHandle_t m_cublasHandle;
+  cudaStream_t m_stream;
+};
+
+/**
+ * @brief Computes batched kronecker product between AkB <- A (x) B
+ * 
+ * @note The block x is the batch id, the thread x is the starting row
+ *       in B and the thread y is the starting column in B
+ * 
+ * @param[in]   A    Pointer to the raw data of matrix `A`
+ * @param[in]   m    Number of rows (A)
+ * @param[in]   n    Number of columns (A)
+ * @param[in]   B    Pointer to the raw data of matrix `B`
+ * @param[in]   p    Number of rows (B)
+ * @param[in]   q    Number of columns (B)
+ * @param[out]  AkB  Pointer to raw data of the result kronecker product
+ * @param[in]   k_m  Number of rows of the result    (m * p)
+ * @param[in]   k_n  Number of columns of the result (n * q)
+ */
+__global__ void kronecker_product_kernel(const double* A, int m, int n, const double* B,
+                                         int p, int q, double* AkB, int k_m,
+                                         int k_n) {
+  const double* A_b = A + blockIdx.x * m * n;
+  const double* B_b = B + blockIdx.x * p * q;
+  double* AkB_b = AkB + blockIdx.x * k_m * k_n;
+
+  for (int ia = 0; ia < n; ia++) {
+    for (int ja = 0; ja < m; ja++) {
```

Comment:
`m` and `n` bounds are inverted (will fix that too in my PR)

Reply:
Solved in https://github.com/rietmann-nv/cuml/pull/3

Reply:
Merged